### ISSUE TITLE
strictCorequisite

### ISSUE THREAD

**Issue by jeannekitchens** [CredentialEngine/Schema-Development#771]
From Fluree team https://curricularanalytics.github.io/CurricularAnalytics.jl/latest/index.html#Terminology-1

Prerequisite : course AA must be completed prior to attempting course BB.
Co-requisite : course AA may be taken prior to or at the same time as attempting course BB.
**Strict co-requisite : course AA must be taken at the same time as course BB.**

This is related to #768 

@siuc-nate @mparsons-ce @jeannekitchens met with Fluree team on June 9th to review #768 and requisites came up as important.  

CTDL terms/definitions:
**https://credreg.net/ctdl/terms#corequisite Credentials that must be pursued concurrently.**
https://credreg.net/ctdl/terms#prerequisite Resource that is required as a prior condition to this resource.

**Of note, we have a wording consistency issue here and should change "Credentials" to Resources.**

**Comment by siuc-nate** [comment]
Actually, I think we already cover this.

Prerequisite: [Already covered](https://credreg.net/ctdl/terms/prerequisite)
(Loose) Corequisite: This is just shorthand for saying two objects require each other (without specificity as to when either is pursued)
(Strict) Corequisite: [Already covered](https://credreg.net/ctdl/terms/corequisite) (Look at the definition, not the label)

**Comment by siuc-nate** [comment]
They didn't specify "prior to or at the same time"; I think that was just Jeanne's interpretation.  Mine is that the student just needs to complete both courses in order for either to count.  We should probably reach out to them to verify that.

**Comment by siuc-nate** [comment]
I missed that Jeanne had updated the original post.  Thanks.

What we're missing here is whether or not `Course AA -[ looseCorequisite ]-> Course BB` is a two-way street or if Course BB serves as a blocker of sorts (in other words, whether my interpretation is correct or whether it's meant to be interpreted as "You can't take Course AA _after_ Course BB but any other timing is fine".  The latter _seems_ to be implied by the definition there, but I want to know for sure from them before we go adding a new property based on that assumption. 

It could well be that `Course AA -[ looseCorequisite ]-> Course BB` actually implies `Course BB -[ looseCorequisite ]-> Course AA`.

**Comment by stuartasutton** [comment]
I don't see why such an assumption is necessary to the utility of the property. If `Course AA -[ looseCorequisite ]-> Course BB` and `Course BB -[ looseCorequisite ]-> Course AA` are operative, then both can assert the relationship and if not only one can so assert.

**Comment by siuc-nate** [comment]
I received confirmation from Greg that the correct interpretation is the second one ("You can't take Course BB unless you have taken or are taking course AA").  I have been trying to think of a means of embodying this with existing CTDL, but I think we would need a new property.  I'm struggling to come up with a name that makes much sense ("Co-Prerequisite", "Soft Prerequisite", "Deferred Prerequisite"?).

Taking a stab at it:
**URI:** ceterms:coPrerequisite
**Label:** Co-Prerequisite
**Definition:** Resource that must be completed prior to, or concurrently pursued during the pursuit of, this resource.
**Comment:** This resource may not be pursued unless the referenced resource is already completed or is being pursued at the same time.

**Comment by stuartasutton** [comment]
Of your options, I think `coPrerequisite` is best...although we can continue to think about it. The definition is a bit off.  Perhaps something slightly more simple: "Resource that must be completed prior to, or at the same time as, this resource." I see no need for the comment.

Why constrain the domain as you have? Couldn't there be similar use cases with a learning opportunity and perhaps even a credential? 

We shouldn't lose track of the need to generalized `corequisite` "Credentials that must be pursued concurrently." Also, `corequisite` has a range of only `ConditionProfile`. Why? The property we are defining here doesn't (and shouldn't) have a range of only `ConditionProfile`.

**Comment by siuc-nate** [comment]
I like "Resource that must be completed prior to, or at the same time as, this resource." but I would clarify the second half a bit:
> "Resource that must be completed prior to, or pursued at the same time as, this resource."

>Why constrain the domain as you have? 

My current goal is to solve this for pathways, but if it makes sense to increase the scope, it may be worth doing now. 

>Couldn't there be similar use cases with a learning opportunity and perhaps even a credential?

Possibly, but should we wait until there is demand for that, since there'll be some interaction with Condition Profile?

>We shouldn't lose track of the need to generalized...

Agreed. I was borrowing the notion (the "verb" if you will) of corequisite from the part of CTDL that has to do with Condition Profile and whatnot, and applying that notion to the context of pathways.  [Prerequisite](https://credreg.net/ctdl/terms/prerequisite), on the other hand, is currently a "pathways only" kind of term that theoretically could also apply in Condition Profile land (though it would probably be redundant with [entryCondition](https://credreg.net/ctdl/terms/entryCondition).  Then there's the difference in approaches between [hasCondition](https://credreg.net/ctdl/terms/hasCondition)/[ComponentCondition](https://credreg.net/ctdl/terms/ComponentCondition)/[requiredNumber](https://credreg.net/ctdl/terms/requiredNumber) and [requires](https://credreg.net/ctdl/terms/requires)/[ConditionProfile](https://credreg.net/ctdl/terms/ConditionProfile)/[alternativeCondition](https://credreg.net/ctdl/terms/alternativeCondition).  There are a lot of overlaps in the venn diagrams for those, but they are distinct enough that we'll probably need to harmonize them at some point. That is beyond the scope of this issue, however.

Anyway, more to your question, we could use coPrerequisite (or whatever we end up calling it) in both areas:
**URI:** ceterms:coPrerequisite
**Label:** Co-Prerequisite
**Definition:** Resource that must be completed prior to, or pursued at the same time as, this resource.
**Domain:** ceterms:PathwayComponent subclasses (same as [ceterms:prerequisite](https://credreg.net/ctdl/terms/prerequisite)), ceterms:Credential + subclasses, ceterms:AssessmentProfile, ceterms:LearningOpportunityProfile
**Range:** PathwayComponent subclasses (same as [ceterms:prerequisite](https://credreg.net/ctdl/terms/prerequisite)), ceterms:ConditionProfile

**Comment by siuc-nate** [comment]
Changes to make:

Create Co-Prerequisite: 
> **URI:** ceterms:coPrerequisite
> **Label:** Co-Prerequisite
> **Definition:** Resource that must be completed prior to, or pursued at the same time as, this resource.
> **Domain:** ceterms:PathwayComponent subclasses (same as [ceterms:prerequisite](https://credreg.net/ctdl/terms/prerequisite)), ceterms:Credential + subclasses, ceterms:AssessmentProfile, ceterms:LearningOpportunityProfile
> **Range:** PathwayComponent subclasses (same as [ceterms:prerequisite](https://credreg.net/ctdl/terms/prerequisite)), ceterms:ConditionProfile, ceterms:Credential + subclasses, ceterms:AssessmentProfile, ceterms:LearningOpportunityProfile

Update [Corequisite](https://credreg.net/ctdl/terms/corequisite):
> **Remove Definition:** Credentials that must be pursued concurrently.
> **Add Definition:** Resources that must be pursued concurrently.
> **Add Domains:** PathwayComponent subclasses (same as [ceterms:prerequisite](https://credreg.net/ctdl/terms/prerequisite))
> **Add Ranges:** PathwayComponent subclasses (same as [ceterms:prerequisite](https://credreg.net/ctdl/terms/prerequisite)), ceterms:Credential + subclasses, ceterms:AssessmentProfile, ceterms:LearningOpportunityProfile

The end result would be that both Corequisite and Co-Prerequisite have the same domain and range.

**Comment by siuc-nate** [comment]
Also consider updating the way that Alternative Condition works to leverage the Required Number property and make "pathways" embodied through condition profiles behave more like "Pathways" do now:

Update the definition for [Alternative Condition](https://credreg.net/ctdl/terms/alternativeCondition):
> **Remove Definition:** Constraints, prerequisites, entry conditions, or requirements in a context where more than one alternative condition or path has been defined and from which any one path fulfills the parent condition.
> **Add Definition:** Constraints, prerequisites, entry conditions, or requirements in a context where more than one alternative condition or path has been defined.

Update [Required Number](https://credreg.net/ctdl/terms/requiredNumber):
> **Add Domain:** ceterms:ConditionProfile

**Comment by stuartasutton** [comment]
This issue currently has a "Pathway Planning" label. We need to be very careful that we don't confuse simple pairwise requirements on resources (e.g., course prerequisite) with our formalization of [ceterms:Pathway](https://purl.org/ctdl/terms/Pathway). To do so would put us on a massively slippery slope.

**Comment by philbarker** [comment]
I removed the Pathway Planning tag. I guess I didn't notice when adding it that it does not relate to pathway components.

**Comment by stuartasutton** [comment]
@siuc-nate, just to be sure that I have this right, in sum, the term definitions are as follows:

> **URI:** ceterms:prerequisite
> **Label:** Prerequisite
> **Definition:** Resource that is required as a prior condition to this resource.

> **URI:** ceterms:coPrerequisite
> **Label:** Co-Prerequisite
> **Definition:** Resource that must be completed prior to, or pursued at the same time as, this resource.

> **URI:** ceterms:coRequisite
> **Label:** Co-Requisite
> **Add Definition:** Resources that must be pursued concurrently.

Nate, I'd like to nail down these definitions with a simple yes/no before looking at domains and ranges. So, yes/no? If no, what explicitly needs correcting.

**Comment by siuc-nate** [comment]
Are the above proposals still valid, given the changes in the June 2022 release for the prerequisite property (which basically emptied out its domain and range)?

**Comment by stuartasutton** [comment]
@siuc-nate & @philbarker, with the June 2022 release, we removed ceterms:prerequisite from the domains of PathwayComponent; however its domain should not have been "emptied out" but changed to ceterms:LearningOpportunity, ceterms:Course.

**Comment by siuc-nate** [comment]
It looks like that got missed in the proposal, then. According to https://github.com/CredentialEngine/Schema-Development/blob/master/PathwaysConditionsAndConstraints/proposal.txt the only actions with prerequisite were to remove its domain and range.

**Comment by stuartasutton** [comment]
Possibly. It was in the long github issue where the changes were developed. I obviously thought it was done (see https://github.com/CredentialEngine/Schema-Development/issues/831#issuecomment-1143996335)

**Comment by jeannekitchens** [comment]
This issue is handled with #846
